search.php

```diff
@@ -14,6 +14,8 @@
  */
 if ( is_main_query() && is_search() ) {
 	if ( 'GET' === filter_input( INPUT_SERVER, 'REQUEST_METHOD' ) ) {
+		global $p4;
```

Comment:
I suppose the introduction of global variable `$p4` has to do with the hooks related to ajax requests.
I am not sure if we can avoid it but can we add the hooks to master theme like below? Will this work for our case?
`add_action( 'wp_ajax_get_paged_posts',        array( 'P4_Search', 'get_paged_posts' ) );`

Reply:
Yes, it is solely related to the ajax requests.
I struggled with this quite a bit. Did not want to use a global either... cause globals are baaaad mkaaaay.... although WP itself uses them quite extensively to make development easier.

I had already tried your suggestion but this did not work in our case, because the get_paged_posts() method can not be static as it needs to access the P4_Search object itself and set/get some of its properties/methods and at that point the already created P4_Search object is not accessible. Thing is I wanted to try it a bit more but somebody whose name begins with K would bite my neck if I delayed more :-)

Your comment though allowed me try it out again, but this time create a new P4_Search object in each ajax request and separate functionality depending whether we are in ajax or normal request. Just tried this and seems to work.
It has a disadvantage of creating a new -local- P4_Search instance (not that big of a disadvantage, we can live with it) but at least we get rid of the global $p4 which is more important.
@kirdia @koyan